### DISCUSSION TITLE
Kafka messages JSON schema

### DISCUSSION THREAD

**Discussion by rnmagalhaes** [ClickHouse/ClickHouse#47933] [Q&A] [answered]
We are using the built-in Kafka engine, together with materialized views, to sink multiple topics to ClickHouse tables.

The format of the Kafka messages is JSON, the schema is somewhat still under development. Once the schema is defined, the idea is to  switch to Confluent's (or RedPandas's) schema registry and all the messages to Avro.

Now to the question:
Being still under development, the JSON schema can be whatever we decide it to be. Assuming we will use it mainly to materialize events to ClickHouse tables, what kind of schema would you guys prefer?
1. Completely flat JSON: this would facilitate the creation of the Kafka table and the associated Materialized View
`{
	"isbn": "123-456-222",
	"author_lastname": "Doe",
	"author_firstname": "Jane",
	"editor_lastname": "Smith",
	"editor_firstname": "Jane",
	"title": "The Ultimate Database Study Guide",
	"category": ["Non-Fiction","Technology"]
}
`
2. Somewhat nested JSON: A bit more complicated to create the Kafka table (using JSONAsString) and MV (using JSONExtractString to parse each nested attribute), but more structured and easy for a human to read.
`{
	"isbn": "123-456-222",
	"author": {
		"lastname": "Doe",
		"firstname": "Jane"
	},
	"editor": {
		"lastname": "Smith",
		"firstname": "Jane"
	},
	"title": "The Ultimate Database Study Guide",
	"category": ["Non-Fiction","Technology"]
}`
3. None of the above...

I already had a look at https://clickhouse.com/docs/en/integrations/data-formats/json. ClickHouse is able to parse most of the formats. My main concern is simplicity, maintenance and stability.

Thanks!

**Comment by mdonkers** [comment, marked as answer]
This depends mostly on how you're planning to query the data, and which parts of the JSON body might still change.
In general ClickHouse performs best with separate (flat) column structure. But if you have (nested) JSON attributes that might change, then storing those as JSON blobs is easier.

If you have bigger data-sets available, it might make sense trying out some different structures and running some performance tests to get an understanding of what works best for your scenarios.

I understand this answer might not be what you're looking for. But it depends too much on how you'll query the data.
